plugins/storage/pyparts/ceph_daemon_checks.py

```diff
@@ -179,10 +188,23 @@ def get_ceph_versions_mismatch(self):
             for version in versions[daemon_type]:
                 vers.append(version)
                 global_vers.add(version)
-
+                if self.version_compare(str(version)) > \
+                        self.version_compare(h_version):
+                    h_version = version
+                    h_daemon = daemon_type
             if vers:
                 daemon_version_info[daemon_type] = vers
 
+        for daemon_type in versions:
+            if daemon_type == 'mon':
+                for version in versions[daemon_type]:
+                    if self.version_compare(str(version)) < \
+                          self.version_compare(h_version):
+                        msg = ("mon version {} is lower than {} version {}"
+                               .format(version, h_daemon, h_version))
+                        issue = issue_types.CephDaemonError(msg)
+                        issue_utils.add_issue(issue)
+
```

Comment:
Does this cover all scenarios where a mismatch occurs? My understanding is that we are looking to identify if one or more daemons has versions that break the ordering required by Ceph [1] and that could apply to more than just mons iiuc.
 
[1] https://access.redhat.com/documentation/en-us/red_hat_ceph_storage/2/html/installation_guide_for_ubuntu/upgrading_ceph_storage_cluster

Reply:
Hi Ed,

Looking at https://github.com/canonical/hotsos/issues/142 the description seems to suggest that we should issue an error if a mon's version is behind any other components version. (including other mons if present)

(For the general check of mismatched packages, i.e all packages not on same version, we already do issue a warning.)

I did see the doc you linked, especially the suggestion to upgrade in the following order,

Monitor nodes
OSD nodes
Ceph Object Gateway nodes
All other Ceph client nodes

So do you mean that we would need to also check if an OSD node is behind an object gateway node, or any other client node (rgw/mds)? , and similarly also check if an object gateway node is behind any other client node except a mon or osd?

Reply:
As discussed, I am leaving the check as is, i.e we will issue an error if a mon is lower than any other cluster component. As discussed, for other components, the existing warning will suffice.